[OGUI-1648] update layout list project structure

NarrowsProjects opened the pull request:
#### I have JIRA issue created
- [x] branch and/or PR name(s) includes JIRA ID
- [x] issue has "Fix version" assigned
- [x] issue "Status" is set to "In review"
- [x] PR labels are selected
- [x] FLP integration tests were ran successful

This PR is a improves the project structure for the layout-list page.

- A new pages module called layoutListView has been made.
- Layout lists have their own model public/pages/layoutListView/model/LayoutListModel.js. 
- Pages/list/page.js has been moved to pages/LayoutListView/component/LayoutListPage.js
- Pages/list/header.js has been moved to pages/LayoutListView/component/LayoutListHeader.js
- Responsibility of managing the folder class has been moved from Model.js to LayoutListModel.js
- Layout card have been abstracted into being their own component. 
- The LayoutListPage.js no longer relies on on Model.js at all. 
- The Folder class has been changed to a list-like structure that can be toggled.
- The Folder class parses json into objects by passing a constructor as an argument.
- The LayoutCard view has gained a model to control its state.

Commit:
feat: move Folder management from Model to LayoutListModel.js

Commit:
chore: move layout/list/(page|header) to LayoutListView/components.

Commit:
fix(LayoutListModel): rename  LayoutListsModel to LayoutListModel

Commit:
fix(LayoutListPage): fix folder toggle function.

Commit:
test(LayoutListPage): add test that checks if the folders are toggelable.

Commit:
feat: create and implement LayoutListComponent

Commit:
fix: remove model from super constructor and initialise model state variable

Commit:
feat(Layout): move search functionallity to LayoutList

Commit:
feat: move toggleOfficial from Layout to LayoutList

Commit:
feat: make layoutListPage independent of Model

Commit:
refractor(LayoutListModel): remove searchinput from Folder initialisations.

Commit:
feat: create FolderModel.js

Commit:
feat(LayoutList): replace Folder class with FolderModel class

Commit:
feat: create FolderComponent.js

Commit:
refractor: replace LayoutListPage implementation with FolderComponent.

Commit:
feat: add layoutCardModel.js

Commit:
feat: isolate LayoutListCard

mariscalromeroalejandro commented on QualityControl/public/pages/layoutListView/components/LayoutListPage.js:
```diff
@@ -0,0 +1,27 @@
+ * @param {object} model - LayoutListModel: respondible for managing the page's state.
+ * @returns {vnode} - virtual node element
+ */
+export default function (model) {
```
In this case, it's preferable to pass the folders directly as arguments rather than the entire model

NarrowsProjects replied:
Understood

mariscalromeroalejandro commented on QualityControl/test/public/pages/layout-list.test.js:
```diff
@@ -56,6 +56,38 @@ export const layoutListPageTests = async (url, page, timeout = 5000, testParent)
+    strictEqual(nrOfOpenedFolders, 1, 'Official Layouts should have opened');
+
+    await page.click(toggleFolderPath(myLayoutIndex)); // This will open a folder
+    await delay(200);
```
too many delays, are all necessary?

NarrowsProjects replied:
clicking the folder toggle sends  a get-request to the backend.
The 200ms is just an arbitrary number though, And I'll see if the time can be reduced.

Furthermore, the delays at 145, 118  and 99 are indeed unneeded.

graduta commented on QualityControl/public/folder/view/FolderComponent.js:
```diff
@@ -0,0 +1,67 @@
+  return h(
+    '.m2.shadow-level3.br3.flex-column',
+    [
+      createHeaderOfFolder(model),
```
I see there is a bit of a lack of consistency in the namings. For example, the folders have 2 components: a header and a body.
But for the function you use `createHeaderOfFolder` and `folderBody`. 
I would recommend you use one convention. As all vnodes in this frameworks are created by functions, we can remove the `create`. Thus, please use `headerOfFolder` or `folderHeader` and similar for the body

graduta commented on QualityControl/public/folder/view/FolderComponent.js:
```diff
@@ -0,0 +1,67 @@
+ * @param {Folder} folder - folder model
+ * @returns {vnode} - virtual node element
+ */
+function createHeaderOfFolder(folder) {
```
I believe it is better to name the variable `folderModel` as to not be confused with the structure of a folder that is represented in the front-end

graduta commented on QualityControl/public/folder/view/FolderComponent.js:
```diff
@@ -0,0 +1,67 @@
+  return h(
+    '.p2.object-selectable',
+    {
+      style: 'border-radius: .5rem .5rem 0 0; display: flex; flex-direction: row',
```
You should avoid defining styles in the component but instead use a `class` in the css file.

graduta commented on QualityControl/public/folder/view/FolderComponent.js:
```diff
@@ -0,0 +1,67 @@
+      onclick: () => folder.toggleFolder(),
+    },
+    [
+      h('b', { style: 'flex-grow:1;' }, [
```
As well as above

graduta commented on QualityControl/public/folder/view/FolderComponent.js:
```diff
@@ -0,0 +1,67 @@
+  return layouts.match({
+    NotAsked: () => null,
+    Loading: () => h('div', 'Loading...'),
+    Failure: () => h('div', [h('div.alert.alert-danger', 'Unable to retrieve this list of layouts')]),
```
I do not see any `alert` and `alert-danger` classes defined anywhere in the CSS set. Why do we need them here?

graduta commented on QualityControl/public/pages/layoutListView/components/LayoutListPage.js:
The page itself is the parent of all components part of `layoutListView`
Thus, it should be outside of this directory so that it is obvious to future developers that it is the main point of entry for this page

graduta commented on QualityControl/public/pages/layoutListView/components/LayoutListCard.js:
```diff
@@ -0,0 +1,96 @@
+/**
```
You are missing the CERN copy right documentation. Please see the other files as example

graduta commented on QualityControl/public/pages/layoutListView/components/LayoutListCard.js:
```diff
@@ -0,0 +1,96 @@
+ * @param {object} model - LayoutCardModel: the model handeling the state of this singular view.
+ * @returns {vnode} Complete layout card virtual DOM node
+ */
+export default function (model) {
```
Here, the model should be named accordingly. `model` should be used only for the root model of the application
In this case, the variable should be called layoutCardModel

graduta commented on QualityControl/public/pages/layoutListView/model/LayoutCardModel.js:
```diff
@@ -0,0 +1,54 @@
+   * @returns {Promise<boolean>} True if user has sufficient authority
+   */
+  async sufficientAuthority() {
+    return this.model.model.model.session.access.some((role) => isUserRoleSufficient(role, UserRole.GLOBAL));
```
I see that above you are passing the `model` - model of the application. 
Thus, `this.model` can be used here instead of the chaining of models you used

graduta commented on QualityControl/public/pages/layoutListView/model/LayoutListModel.js:
```diff
@@ -0,0 +1,68 @@
+import { UserRole, isUserRoleSufficient } from './../../../library/userRole.enum.js';
```
Missing copyright docs

graduta commented on QualityControl/public/pages/layoutListView/model/LayoutListModel.js:
```diff
@@ -0,0 +1,68 @@
+import FolderModel, { FolderType } from '../../../folder/model/FolderModel.js';
+import LayoutCardModel from './LayoutCardModel.js';
+
+export default class LayoutListModel extends Observable {
```
Missing class documentation

graduta commented on QualityControl/public/pages/layoutListView/model/LayoutListModel.js:
```diff
@@ -0,0 +1,68 @@
+    this._initializeFolders();
+  }
+
+  _initializeFolders() {
```
Missing function documentation

graduta commented on QualityControl/public/pages/layoutListView/model/LayoutListModel.js:
```diff
@@ -0,0 +1,68 @@
+  /**
+   * Given an ID and new value for official status, update it accordingly
+   * @param {string} id - of layout to modify
+   * @returns {void}
```
The returns should be a convention as well. Above you used `undefiend` (as majority of the codebase uses), while here you used `void`.
Please keep the most used one only

graduta reviewed (changes requested):
Very good start with the refactoring! There are a few things missing that I added in the review

Commit:
test: remove unneeded delays from tests

Commit:
test: reduce delay times.

The working delay times have been reduced to the minimum required time. By running the test 10 times without fail.

Commit:
refactor: rename createHeaderOfFolder to folderHeader

Commit:
refactor: pass layoutList.folder to layoutlistpage instead of the full model

Commit:
refractor(FolderComponent): remove style declaration

Commit:
refractor(FolderComponent): rename folder/model to folderModel

Commit:
style(folderComponent): replace non-existent class

Commit:
refractor(LayoutListCard): rename model to layoutCardModel

Commit:
refractor: more efficiently pass parent model to LayoutCardModel.

Commit:
chore: add copyright licensing to files

Commit:
chore(LayoutListModel): add missing documentation

Commit:
fix(LayoutCardModel): remove async annotation from function that isn't async

Commit:
chore: move LayoutListPage to parent directory

Commit:
feat: add repository function for fetching layoutCards.

Commit:
feat: add repository function for fetching layoutCards.

Commit:
feat: add layoutController function for fetching layoutCards.

Commit:
feat: add layoutController function for fetching layoutCards.

Commit:
feat: change layout layout.service.js GET endpoint

Commit:
feat: change layout layout.service.js GET endpoint

Commit:
feat: add getLayoutCardsByUserId function to Layout.service.js

Commit:
feat: implement new layout.service.js endpoint

Commit:
test: add layoutController and Layout repository tests for layoutCards

Commit:
fix: remove unneeded fetch from layoutservice.js

Commit:
refractor(LayoutRepository): change listLayoutcards to use json deconstruction to make the filtertype more apparent.

Commit:
refractor(LayoutRepository): change listLayout to determine the fields it returns based on a fields array argument

Commit:
feat(LayoutController): improve error handeling

Commit:
feat(LayoutService): changed getLayoutCards(byUserId) to use the layouts endpoint with the fields query parameter

Commit:
refractor: remove unused Layoutcard REST functions.

Commit:
feat(layoutDto): add LayoutsGetDto

Commit:
feat: implement new LayoutGetDTO

Commit:
refactor(LayoutRepository): rename filter arrow function to LayoutFilter.

Commit:
refactor: replace getLayoutCards(ByUserId) with getLayout(ByUserId)

Commit:
refactor(LayoutRepository::ListLayouts): remove unneeded Array-check

Commit:
refractor: log a warning if a field is missing instead of throwing an error

Commit:
fix(LayoutController): incorrect fields pass to repository

Commit:
refractor(LayoutRepository): remove unneeded loop from ListLayouts

Commit:
feat: add RequestField enum

Commit:
refractor: use RequestFields.LAYOUT_CARD in Model and Layout

Commit:
refractor(Layout.service): abstract comparitor and implement it in layout sort functions

Commit:
feat(LayoutCardModel): add equals function

Commit:
feat(FolderModel): add Push and RemoveItem functions

Commit:
feat(LayoutListModel): add removeLayoutFrom and AddLayoutTo

Commit:
feat(LayoutCardModel): remove getLayouts and getLayoutsByUserId

Commit:
chore: update documentation

Commit:
chore: update documentation

Commit:
test: add more tests for coverage

Commit:
chore: specify loggers for LayoutController/Reporitory

Commit:
refractor(LayoutController): split error handeling up into two try-catches

Commit:
Merge branch 'feature/QCG/OGUI-1655/update-layout-list-rest-to-only-send-required-fields' of github.com:AliceO2Group/WebUi into feature/QCG/OGUI-1648/update-layout-list-project-structure

mariscalromeroalejandro commented on QualityControl/lib/controllers/LayoutController.js:
```diff
@@ -56,15 +59,32 @@ export class LayoutController {
+      const layouts = await this._layoutRepository.listLayouts({ filter, fields });
+      return res.status(200).json(layouts);
+    } catch (error) {
+      logger.debugMessage(`Error retrieving layouts: ${error}`);
```
why debug if it's an error message?

NarrowsProjects replied:
My logic was that this message will only be used for debugging, but now that you mention it, it can just as easily be done with errorMessage

mariscalromeroalejandro commented on QualityControl/lib/repositories/LayoutRepository.js:
```diff
@@ -11,22 +11,51 @@
-    return this._jsonFileService.data.layouts.filter((layout) =>
-      (filter.owner_id === undefined || layout.owner_id === filter.owner_id)
-            && (filter.name === undefined || layout.name === filter.name));
+  listLayouts({ name, owner_id, fields } = {}) {
```
Default fields to an empty array if it's not passed in
```suggestion
  listLayouts({ name, owner_id, fields = [] } = {}) {
```

Commit:
fix: pass filter between  layoutController and layoutRepository

Commit:
fix: pass filter between  layoutController and layoutRepository

NarrowsProjects committed:
Merge branch 'dev' into feature/QCG/OGUI-1655/update-layout-list-rest-to-only-send-required-fields

Commit:
Merge remote-tracking branch 'origin' into feature/QCG/OGUI-1648/update-layout-list-project-structure

mariscalromeroalejandro commented on QualityControl/public/services/Layout.service.js:
```diff
@@ -150,7 +177,7 @@ export default class LayoutService {
   async patchLayout(id, patch) {
     try {
-      return RemoteData(await jsonPatch(`/api/layout/${id}`, { body: { ...patch } }));
+      return RemoteData.success(await jsonPatch(`/api/layout/${id}`, { body: { ...patch } }));
```
I think assigning await... to a constant `response` could improve the error handling, also readability

NarrowsProjects replied:
Agreed

Commit:
refractor(LayoutRepository): set fields to array by default

Commit:
refactor(LayoutService):  save patch response in a variable

NarrowsProjects committed:
Merge branch 'dev' into feature/QCG/OGUI-1648/update-layout-list-project-structure

Commit:
chore: replace debug logs with error logs

Commit:
chore: add/fix documentation for missing classes

Commit:
refactor: remove unused ToggleOficial

Commit:
fix(LayoutDto): make fields optional to prevent errors when

Commit:
test: add api-layout-get test

Commit:
feat(Layoutcontroller): improve errorHandeling for getLayoutHandler

Commit:
Merge branch 'feature/QCG/OGUI-1655/update-layout-list-rest-to-only-send-required-fields' of github.com:AliceO2Group/WebUi into feature/QCG/OGUI-1648/update-layout-list-project-structure

Commit:
chore: remove Folder.js

Commit:
Merge branch 'dev' of github.com:AliceO2Group/WebUi into feature/QCG/OGUI-1648/update-layout-list-project-structure

mariscalromeroalejandro committed:
fix lint error

Commit:
test: add aditional test to test for getLayout (rundefinition + pdpBeamType)

Commit:
Merge branch 'dev' of github.com:AliceO2Group/WebUi into feature/QCG/OGUI-1648/update-layout-list-project-structure

Commit:
refactor: remove unneeded fetch from save call.

Commit:
fix: remove name-filter from getLayoutsHandler.

I was informed by Alejandro that the database will have the name column unique.
Which means that filtering by name serves no purpose anymore as it will be identical to getLayoutByNameHandler

Commit:
git commit -m "test: remove name filter tests for listLayouts and getLayoutsHandler"

mariscalromeroalejandro commented on QualityControl/lib/repositories/LayoutRepository.js:
```diff
@@ -28,7 +28,7 @@ export class LayoutRepository extends BaseRepository {
    * @throws {Error} If any specified field does not exist in the layout objects
    */
-  listLayouts({ name, owner_id, fields } = {}) {
+  listLayouts({ name, owner_id, fields = [] } = {}) {
```
You don't need to check again in line 40 if `!fields` because you are setting a default value to `[]`

NarrowsProjects replied:
Yes, but in LayoutController, it can be provided as undefined if fields isn't provided. Meaning that it can well be undefined

NarrowsProjects replied:
Never mind, you were right. I love JavaScript

Commit:
refactor: replace unrequired truthy check for ? chaining

Commit:
refractor(LayoutRepository): remove unneeded ? chain

Commit:
refactor(LayoutList): project breadth refactor

Commit:
refactor: replace unneeded arrow functions

Commit:
refactor: remove layoutHeader model dependency

mariscalromeroalejandro commented on QualityControl/public/folder/view/FolderComponent.js:
```diff
@@ -4,34 +4,40 @@ import LayoutListCard from '../../pages/layoutListView/components/LayoutListCard
+ * @returns {vnode} - Virtual DOM node representing the folder header
  */
-function folderHeader(folderModel) {
+function folderHeader({ folderType, isOpened, title }, clickFunction) {
```
much better!

mariscalromeroalejandro commented on QualityControl/public/pages/layoutListView/components/LayoutListCard.js:
```diff
@@ -29,43 +29,34 @@ import { iconBadge } from '/js/src/icons.js';
-        onclick: (e) => model.router.handleLinkEvent(e),
-      }, name),
-    ]),
+    h('h5', [h(`a.${textColor}`, { href, onclick: clickHandeler }, name)]),
```
Hendeler -> Handler

mariscalromeroalejandro commented on QualityControl/public/layout/Layout.js:
```diff
@@ -808,4 +808,14 @@ export default class Layout extends Observable {
+  /**
+   * Determines whether the current authenticated user owns the specified layout.
+   * Compares the current session user's person ID with the owner ID of the given layout item to verify ownership.
+   * @param {object} layoutOwnerId - The layout item to check ownership against. Must contain an `owner_id` property.
```
This is ambiguous and likely incorrect, it suggests layoutOwnerId is an object with an owner_id property, but the code compares directly to layoutOwnerId

NarrowsProjects replied:
It is indeed incorrect, just a remnant of what i wrote before.

mariscalromeroalejandro commented on QualityControl/public/layout/view/header.js:
```diff
@@ -20,25 +20,26 @@ import {
  */
-export default (model) => model.layout.item
-  ? model.layout.editEnabled ? toolbarEditMode(model) : toolbarViewMode(model)
+export default (layout) => layout.item
```
reduce nesting checking first `if (!layout.item) return null` (readability)

NarrowsProjects replied:
```js
export default (layout) => layout.item && layout.editEnabled ? toolbarEditMode(layout) : toolbarViewMode(layout);
 ```
Something like this?

mariscalromeroalejandro commented on QualityControl/public/layout/view/header.js:
```diff
@@ -20,25 +20,26 @@ import {
-const toolbarViewMode = (model) => {
-  const layoutItem = model.layout.item;
-  const { isOfficial } = layoutItem;
+const toolbarViewMode = (layout) => {
```
this method has 53 lines and I think that, now that we are refactoring the structure, it is a good moment to separate this function into several smaller functions like rendering the button group (after the  `// Show group button edit/duplicate only for owner of the layout shown` comment)

mariscalromeroalejandro commented on QualityControl/public/layout/view/header.js:
```diff
@@ -51,33 +52,33 @@ const toolbarViewMode = (model) => {
         }, iconShareBoxed()),
-        model.session.personid == layoutItem.owner_id && [
+        layout.ownsLayout(owner_id) && [
           h('.dropdown', {
```
you can also put this into a `renderOwnerControls` function available only to the layout owner. so that later un the render buttongroup you can add `isOwner && renderOwnerControls(layout)` and have a nice an clear structure

NarrowsProjects replied:
I abstracted all buttons that have logic in them into their own functions, meaning that it looks like this:
```js        
layout.ownsLayout(owner_id) && [editDropdown(layout), deleteButton(layout)],
```

mariscalromeroalejandro commented on QualityControl/public/layout/view/header.js:
```diff
@@ -110,21 +111,21 @@ const toolbarViewModeTab = (model, tab, i) => {
   h('.w-50.text-center', [
     h('div', { class: 'header-layout' }, [
-      h('span', model.layout.item.tabs.map((tab, i) => toolbarEditModeTab(model, tab, i))),
+      h('span', layout.item.tabs.map((tab, i) => toolbarEditModeTab(layout, tab, i))),
```
I would move also this to a new function called for example `createTabButtons` (readability)

NarrowsProjects replied:
I implemented it but named the function editTabLinks

mariscalromeroalejandro commented on QualityControl/public/layout/view/header.js:
```diff
@@ -110,21 +111,21 @@ const toolbarViewModeTab = (model, tab, i) => {
         tabBtn({
           title: 'Add new tab to this layout',
           class: 'default',
           onclick: () => {
```
and this one to `handleAddTab`

NarrowsProjects replied:
I implemented it but named the function viewTabLinks

mariscalromeroalejandro commented on QualityControl/public/layout/view/header.js:
```diff
@@ -134,18 +135,18 @@ const toolbarEditMode = (model) => [
       type: 'text',
-      value: model.layout.item.name,
+      value: layout.item.name,
       oninput: (e) => {
```
same here `handleInputChange`

mariscalromeroalejandro commented on QualityControl/public/layout/view/header.js:
```diff
@@ -203,22 +204,17 @@ const toolbarEditModeTab = (model, tab, i) => {
-  ]);
+    onchange: (e) => layout.resizeGridByXY(e.target.value),
+  }, [1, 2, 3, 4, 5].map((i) =>
+    h('option', { selected: tab?.columns === i, title: `Resize layout to ${i} columns`, value: 1 }, `${i} cols`)));
```
beautiful refactorization

Commit:
refactor: feedback 1

Commit:
refactor(filter.js): refactor layoutFilter

graduta commented on QualityControl/public/pages/layoutListView/model/LayoutListModel.js:
```diff
@@ -0,0 +1,105 @@
+import LayoutCardModel from './LayoutCardModel.js';
+
+/**
+ * Model namespace with all requests to load or create layouts, compute their position on a grid,
```
Probably doc was copied from previous class as it does not correctly describe this model?

NarrowsProjects replied:
Indeed

graduta commented on QualityControl/public/pages/layoutListView/model/LayoutListModel.js:
```diff
@@ -0,0 +1,105 @@
+  constructor(model) {
+    super();
+    this.model = model;
+    this.searchInput = '';
```
Now that we refactor, perhaps this can be used as a private variable with a getter and setter exposed?

graduta commented on QualityControl/public/pages/layoutListView/model/LayoutListModel.js:
```diff
@@ -0,0 +1,105 @@
+   * @param {LayoutCardModel} layout - Layout instance to add
+   * @returns {void}
+   */
+  addLayoutTo(folderName, layout) {
```
the name of the method suggests the order of the parameter should be the other way around:
`addLayoutTo(layout, folderName)`
Moreover, should we not check that folderName exists before trying to push? Otherwise you will have an undefined error

NarrowsProjects replied:
I've made a rather unorthodox change that allows you to call the function like so addLayout(layout).to(folderName)

graduta commented on QualityControl/public/pages/layoutListView/model/LayoutListModel.js:
```diff
@@ -0,0 +1,105 @@
+   * @param {LayoutCardModel} layout - Layout instance to remove
+   * @returns {void}
+   */
+  removeLayoutFrom(folderName, layout) {
```
The same comments as for the `addLayoutTo` applies

NarrowsProjects replied:
ditto

graduta commented on QualityControl/public/pages/layoutListView/model/LayoutCardModel.js:
```diff
@@ -0,0 +1,95 @@
+   * Creates a new LayoutCardModel instance
+   * @param {Model} model - The parent model utilizing this card
+   * @param {object} layout - The layout data object containing:
+   * @param {string} id - Unique identifier for the layout
```
The standard doc format for objects within another object is:
```jsdoc
* @param {object} layout - The layout data object containing:
* @param {string} layout.id - Unique identifier for the layout
```
Please update the docs accordingly for all the fields

graduta commented on QualityControl/public/pages/layoutListView/components/LayoutListCard.js:
```diff
@@ -0,0 +1,101 @@
+ * or submit itself to any jurisdiction.
+ */
+
+/**
```
Where is the doc below coming from and why it is placed above an import

NarrowsProjects replied:
A remnant from a previous version I've read past.

graduta commented on QualityControl/public/pages/layoutListView/model/LayoutCardModel.js:
```diff
@@ -0,0 +1,95 @@
+   * @async
+   * @returns {boolean} True if user has sufficient authority
+   */
+  sufficientAuthority() {
```
This method is defined twice identical:
- LayoutCardModel
- LayoutListModel

What about extractinc the method in a utility function that expects the model (or a list of accesses) and the role to check against and use it straight from there?

Commit:
refactor(LayoutListModel): rework the namespace documentation

Commit:
fix: check for folder existence before setting/removing cards

Commit:
refactor: change the call syntax on add/removeLayoutFrom/to

Commit:
chore: update LayoutCardModel constructor jsdoc

Commit:
remove jsdoc remnant from layoutListCard

Commit:
remove sufficientAuthoity from layoutListModel

Commit:
refactor abstract sufficientAuthority to a util class

Commit:
remove commented code

NarrowsProjects committed:
Merge branch 'dev' into feature/QCG/OGUI-1648/update-layout-list-project-structure
